### DISCUSSION TITLE
{ ...blockProps } always on the parent element can make block creation more difficult

### DISCUSSION THREAD

**Discussion by cbravobernal** [WordPress/gutenberg#36079] [Developer Experience]
## What problem does this address?

**TL:DR:**
Block attributes are mostly applied to a wrapper element. What would be the best approach if we need to have some of them on the wrapper element and some of them in the inner element?

Right now, the entire approach of styling blocks is done on a wrapper HTML element. Some examples:

Image block:
![Screenshot 2021-10-28 at 17 28 36](https://user-images.githubusercontent.com/37012961/139287608-55f88bc7-c071-41e3-b381-944e6789dc8c.png)

Post Excerpt Block:
![Screenshot 2021-10-28 at 17 30 30](https://user-images.githubusercontent.com/37012961/139288074-2265208b-1599-487c-a2a5-770c433c7dd0.png)

Using this approach, we get some **benefits** like:
 - Easy to select the block
 - Visual guide of the block content
![Screenshot 2021-10-28 at 17 38 32](https://user-images.githubusercontent.com/37012961/139289365-ad98cc8d-3f0d-4891-a9be-bee4ba52da5e.png)

Sometimes, an element style must be applied to the inner tag. In order to solve it, we use CSS inherit properties:
![139290857-d82cf22d-8997-4046-ba96-346b5e026322](https://user-images.githubusercontent.com/37012961/139425422-aba15eac-b3dd-4610-b8d7-80bd3e3e1555.png)

That is a good approach for many properties like margin, padding, width, height or border-radius.

But what if we want a property to be applied to the inner HTML tag? The main one? Consider we want to achieve this:
![Screenshot 2021-10-29 at 12 58 05](https://user-images.githubusercontent.com/37012961/139423299-3134e596-05ff-40f8-a98e-b550a8903ca6.png)

If we try to use the options provided by the block, this is what we get:
![Screenshot 2021-10-29 at 13 04 00](https://user-images.githubusercontent.com/37012961/139424034-696e6d12-ed99-4760-99a5-81e208a5a91a.png)

If we apply the `{...blockProps}` only to the inner tag. Some attributes like `padding` or `margin` need some tweaks in order to work as expected.
If we apply the `{...blockProps}` in both the wrapper and the main tag. Anything can happen, depending on the block. Like duplicated styles, ignored styles, etc. Not getting the desired solution both in the editor and in frontend.

Of course, all of this can be solved by applying custom CSS, but we lose some part of the power of full site editing.

## What is your proposed solution?

Feel free to **help us find a good answer** to this! Blocks are evolving, having more options, and many of them are being hard to apply due to this issue.

Could we find any way to separate between "Container Attributes" and "Element Attributes"?

Related discussions:

- https://github.com/WordPress/gutenberg/pull/35396#pullrequestreview-790454482
- https://github.com/WordPress/gutenberg/pull/35396#discussion_r723367012
- https://github.com/WordPress/gutenberg/pull/35396#issuecomment-948524448

**Comment by luisherranz** [comment]
Hey @c4rl0sbr4v0, I have started a conversation about the same topic in https://github.com/WordPress/gutenberg/discussions/35473, in case you want to follow it 🙂

For your specific problem: have you tried using the experimental API that @oandregal introduced in https://github.com/WordPress/gutenberg/issues/28913?

I also saw that you opened a related Trac ticket: https://core.trac.wordpress.org/ticket/54319. Could you please explain a bit more the relation of that ticket with this issue?

**Reply by cbravobernal** [reply]
> Hey @c4rl0sbr4v0, I have started a conversation about the same topic in #35473, in case you want to follow it 🙂

Awesome! is there any way to "merge" the discussions?


> For your specific problem: have you tried using the experimental API that @oandregal introduced in #28913?

Didn't use skip serialization and styles, but maybe could solve the issue. I will do some tests!


> I also saw that you opened a related Trac ticket: https://core.trac.wordpress.org/ticket/54319. Could you please explain a bit more the relation of that ticket with this issue?


They are kind of different issues. 

This one is related to what you said in the other one, to find a way to be able to set where do we want the attributes to be applied, on the wrapper or on other elements of the block.

[The other one ](https://core.trac.wordpress.org/ticket/54319) is related more to PHP rendering. In that ticket, the main issue is that we need a way to separate classes from styles. Right now is all in one string, and some WordPress functions have different variables for classes and other attributes.

**Reply by luisherranz** [reply]
> In that ticket, the main issue is that we need a way to separate classes from styles. Right now is all in one string, and some WordPress functions have different variables for classes and other attributes.

I see, thanks.

In any case, I guess some dynamic blocks may also need to distribute block supports features more freely between HTML markup of the block as well.

I still need to study the current APIs more to get a better idea of how it works, though.

> Didn't use skip serialization and styles, but maybe could solve the issue. I will do some tests!

Could you please report here once you do? Thanks Carlos! 🙂

**Comment by cbravobernal** [comment]
We created some PHP code in order to achieve the same we do with `skipSerialization` properties and then `useSpacingProps`.

Maybe we could move it to the Core as a function and make it available for other blocks: what do you think?

```php
$spacing_attributes = isset( $attributes['style']['spacing'] ) ? $attributes['style']['spacing'] : null;
	if ( isset( $spacing_attributes ) && ! empty( $spacing_attributes ) ) {
		$spacing_array = array();
		foreach ( $spacing_attributes as $spacing_attribute_key => $spacing_attribute_value ) {
			foreach ( $spacing_attribute_value as $position_key => $position_value ) {
				$spacing_array[] = $spacing_attribute_key . '-' . $position_key . ': ' . $position_value;
			}
	         }
$spacing_string = implode( ';', $spacing_array );
	}
.....
....
....
if ( isset( $spacing_attributes ) ) {
	return sprintf( '<div style="%1s">%2s</div>', $spacing_string, $avatar_block );
}
```

**Reply by luisherranz** [reply]
I'll put this on hold for now, as we still need to figure out how to modify the current block supports system to make it compatible with declarative frontend hydration.

**Comment by gziolo** [comment]
There is a discussion happening on how to improve the saving and rendering of block styles happening in https://github.com/WordPress/gutenberg/discussions/37495. As a result, @andrewserong opened an issue https://github.com/WordPress/gutenberg/issues/38167 that proposes a new Styles Engine. For this particular case, the solution might be completely different while such a system is in place. At the moment the biggest limitation is that styles are injected into the markup generated inside the wrapper. That's why developers use `__experimentalSkipSerialization` flag to ensure that styles/class names aren't injected in the wrapper if they want to target the inner element. However, as soon as we transition to the world where blocks have some unique class names applied to the wrapper, and the CSS styles are generated on the fly during frontend rendering for them based on block supports, then we no longer have this limitation. We can generate CSS styles that can target any nested element that is separate from the block markup.

**Before**

HTML for the block in the content:
```html
<div class="my-block" style="color:red">
    <div class="my-block__nested" style="color:green">Content</div>
</div>
```

**After**

HTML for the block in the content:
```html
<div class="my-block my-block-abc123">
    <div class="my-block__nested">Content</div>
</div>
```

Dynamically generated CSS styles for the block in the HTML:
```html
<style>
.my-block-abc123 {
    color:red;
}
.my-block-abc123 .my-block__nested {
    color:green;
}
</style>
```

**Comment by cbravobernal** [comment]
After two years and with the styles engine. We can close this stalled discussion. Thanks everyone for your support!
